### DISCUSSION TITLE
Integration of DeepL API for Automatic Field Translation

### DISCUSSION THREAD

**Discussion by stephenhenckaerts** [payloadcms/payload#4370] [Feature Requests & Ideas]
## Feature Request: Integration of DeepL API for Automatic Field Translation

Hello Payload CMS Team,

I am writing to suggest an enhancement that I believe would significantly benefit many users of Payload CMS. As we operate in a global digital environment, the ability to present content in multiple languages is increasingly important. To this end, I propose the integration of the DeepL API for automatic translation of fields.

### Feature Description
The idea is to allow Payload CMS to automatically translate content fields from a source language into multiple target languages using the DeepL API. This integration would enable seamless, high-quality translations, aiding users in creating multilingual content more efficiently.

### Key Benefits
- **Enhanced Multilingual Support:** Automatically translates content, making it easier to manage websites or applications in multiple languages.
- **Time and Resource Efficiency:** Reduces the need for manual translation, saving time and resources.
- **Quality Translations:** DeepL is known for its high-quality translations, which would ensure that the content remains accurate and contextually relevant in different languages.

### Possible Implementation
- Integration could be as an optional plugin or module within Payload CMS.
- Provide configuration options to select source and target languages.
- Allow for manual review and edits of translated content to ensure accuracy.

I believe this feature would be a valuable addition to Payload CMS, particularly for users looking to expand their reach to a global audience. I am excited to hear your thoughts on this proposal and how we might work towards implementing it.

Thank you for considering this feature request.

Best regards,
Stephen

**Comment by matthijs166** [comment]
Have you already seen this plugin? https://github.com/r1tsuu/payload-plugin-translator

Maybe you can make a pull request to the plugin, to add DeepL.
You only need to edit the following function:

https://github.com/r1tsuu/payload-plugin-translator/blob/25c3de9c64a9dd8ecd9df8213ab992356d48b1f7/src/translator/services/google.ts#L9-L29

ChatGPT is also a big help to transform the function:
```ts
export const createDeepLTranslateService = (authKey: string) => {
  return async (
    text: string,
    targetLang: string,
    sourceLang?: string,
    context?: string
  ): Promise<string> => {
    const apiUrl = 'https://api-free.deepl.com/v2/translate';

    const requestBody = {
      text: [text],
      target_lang: targetLang,
      source_lang: sourceLang,
      context: context,
    };

    const response = await fetch(apiUrl, {
      method: 'POST',
      body: JSON.stringify(requestBody),
      headers: {
        'Authorization': `DeepL-Auth-Key ${authKey}`,
        'Content-Type': 'application/json',
      },
    });

    const { translations } = await response.json();

    return translations[0].text as string;
  };
};
```

**Comment by matthijs166** [comment]
@stephenhenckaerts apparently you can pass your own translation service in the plugin 🥳 

https://github.com/r1tsuu/payload-plugin-translator/issues/4#issuecomment-1847078662
